### ISSUE TITLE
Add support for fuzzy_search() in Groonga

### ISSUE THREAD

**Issue by kou** [pgroonga/pgroonga#100]
See also #40.

**Comment by ghevge** [comment]
Any progress on this one?

**Comment by kou** [comment]
No progress. :<
Do you have an use case?

**Comment by askdkc** [comment]
> This works for the languages where I can split the words, but how to I do this for Japanese for example ?

Well, PostgreSQL supports UTF-8 by default so that you can just put any language keywords (in this case Japanese) like this, and it just works:
 
``` sql
SELECT *
FROM your_table
WHERE column1 &@~ 'こんにちは 日本'
   OR column2 &@~ 'こんにちは 日本'
   OR column3 &@~ 'こんにちは 日本';
```

**Comment by ghevge** [comment]
I've ended up using the &@* instead as it is more closer to what I need.
BTW for this operator to work properly with  different languages to I have to configure those languages for full text support at postgres level or pgroonga is able to handle any language out of the box ?

**Comment by kou** [comment]
`&@*` will not work well with the default tokenizer (`TokenBigram`) for non ASCII based languages such as Japanese.
See also: https://pgroonga.github.io/reference/operators/similar-search-v2.html#for-japanese

**Comment by ghevge** [comment]
&@~ doesn't seem to work well with English either. I just need a way to work OKish with any language.

**Comment by ghevge** [comment]
I have a single DB with single schema which can host string from any languages

**Comment by ghevge** [comment]
Will it work if I will define a set of indexes for Japanese in parallel with the ones for  ASCII based languages ?

**Comment by ghevge** [comment]
Like on the same table to have an index for Japanese and one for ASCII based languages.

**Comment by askdkc** [comment]
> &@~ doesn't seem to work well with English either. I just need a way to work OKish with any language.

Sounds like you need to change both of your operators and indexes depending on the language user use for searching. Interesting problem, you have.

I wonder if creating multiple indexes works in this case.

Is it possible to know which language user use before searching so that you can swap the operators for better search results?

**Comment by ghevge** [comment]
Sometimes I will know but most of the time I wont know which language user is using.

**Comment by kou** [comment]
Could you open a new issue or new discussion for the topic?
It's not related to `fuzzy_search()`.

**Comment by ghevge** [comment]
Is there any functionality available in pgroonga at this moment that mimics fuzzy search? Basically what I want is to have a similar behavior that Google has, when a user is typing incorectly a search string:
eg: if the user types: "imagws" the search query should return all the results containing "images"

**Comment by askdkc** [comment]
@ghevge 

Usually you get your user input from frontend side so that it would be easier to handle typo checking in your input fields using JavaScript (before it hits db to query that keyword). 

Perhaps using typo-js or similar packages would help?

https://www.npmjs.com/package/typo-js

**Comment by ghevge** [comment]
My problem though is that my apps currently supports 40+ languages and I'm planning to add support for more. This means that a user can type at any time a search strig in any of these languages. And they can type not just simple words but phrases too.

**Comment by askdkc** [comment]
@ghevge 

Computer needs dictionaries in order to figure out that inputs has typo or not.

So if you want to support many languages, you need to provide dictionaries for each.

Some languages like Japanese sentence is not word separated like English is. In that case, creating dictionary is not enough to figure out which word is a typo or not.

So I guess it's really tough problem to solve if you try to support many languages.

**Comment by askdkc** [comment]
BTW, meilisearch has Typo tolerant feature.

https://www.meilisearch.com/docs/learn/what_is_meilisearch/comparison_to_alternatives

It works well in English when I tried (some weird behavior in Japanese though). 

If it gets better in other languages in near future, it might be a good solution.

**Comment by kou** [comment]
In general, you need to have correct words to implement the auto type correction.
I think that Google has them.

We can collect correct words by using word based tokenizer such as `TokenNgram` for ASCII based space separated word languages such as English and `TokenMecab` for Japanese. I think that there is no multilingual tokenizer in the world yet. So we can't use the same data for all languages for now.
We need correct words for each language and need to detect language from inputs (data in database and query from user).

We can implement Meilisearch like auto typo correction search feature to Groonga. Because Groonga already have features such as edit distance based fuzzy search to implement the feature.

**Comment by askdkc** [comment]
> We can implement Meilisearch like auto typo correction search feature to Groonga. Because Groonga already have features such as edit distance based fuzzy search to implement the feature.

Can't wait!! 😍

**Comment by ghevge** [comment]
@kou what if you asume the existing DB conted as the "correct words" set?  Fuzzy search doesn't have to be perfect. If 50%+ of its results are close to what user would look for, it should be good enough.  Don't try to come up with something perfect , right from the begining. You know the saying: "Perfect is the enemy of good." ?

Once fuzzy search will be made available in Groonga, I personaly don't see why the Postgres comunity will want to use any other Postgres string search mechanism.

**Comment by askdkc** [comment]
@ghevge 

I might be wrong but after I checked the MeiliSearch code, they seems to use the Levenshtein distance algorithm as one of the methods to handle typo checking.

https://github.com/meilisearch/meilisearch/blob/661d1f90dcc11aa7cea599cafdeb7a0dcd497855/milli/src/search/new/query_term/mod.rs#L484-L505

The Levenshtein distance measures the minimum number of operations (insertions, deletions, or substitutions) required to transform one string into another. By comparing the Levenshtein distance between the user’s query and the indexed terms, MeiliSearch can determine potential typos and provide relevant search results.

But this might work well with alphabetical languages, it probably does not works well with Chinese, Japanese and Korean(CJK).

If you are ok with 50%, try MeiliSearch and tell us what you think. If you like it, then implementing the same thing into Groonga provides nice typo tolerant features. If you don't like it then perhaps we need to look for something else.

**Comment by ghevge** [comment]
MeiliSearch fuzzy search functionality looks good enough, to me.

**Comment by askdkc** [comment]
@ghevge 

Yep. They are good.
It seems they have some issues figuring out Kanji character from the inputs are either Chinese or Japanese though.

https://github.com/meilisearch/meilisearch/pull/3812


They use tokenize library Charabia,  BTW.

https://github.com/meilisearch/charabia

So if Groonga gets similar tokenizers implemented, it probably works well in many languages.

**Comment by ghevge** [comment]
Hey Guys! Do you have an ETA for this functionality? Thanks!

**Comment by kou** [comment]
No ETA for now.
Could you share you use-case?

**Comment by ghevge** [comment]
@kou I've shared it already above.

**Comment by askdkc** [comment]
@ghevge 

The quickest way is creating it by yourself and send a PR😉

That's how OSS works.

**Comment by kou** [comment]
> I've shared it already above.

Oh, sorry.

> Any way to seed up the prioritization of this functionality ?

1. You can send a PR by yourself as askdkc says
2. You can do a financial contribution like Supabase: https://pgroonga.github.io/users/#supabase
    (If you're interested in this, I can explain this more.)

**Comment by ghevge** [comment]
@kou regarding option #2 how much will be to get this feature?

**Comment by kou** [comment]
@ghevge OK. Could you e-mail us (`info@clear-code.com`) for this?

**Comment by abetomo** [comment]
`fuzzy_max_distance_ratio` option of `pgroonga_condition()` is available. 
https://pgroonga.github.io/reference/functions/pgroonga-condition.html
